### ISSUE TITLE
resolve `frc46_token` ambiguity

### ISSUE THREAD

**Issue by LesnyRumcajs** [ChainSafe/fil-actor-states#245]
Why does it have the _official_ token whilst in the other datacap states we have a copy-pasted one? Can we always use the crate?

_Originally posted by @LesnyRumcajs in https://github.com/ChainSafe/fil-actor-states/pull/244#discussion_r1478366611_

**Comment by ruseinov** [comment]
I'm not sure why we did have the copy-pasted one in the first place, but the datacap token is modified to cater for Lotus RPC compat.
What needs to be done is:
1. Use the `crate` instead.
2. Introduce TokenExt in `forest` to cater for RPC compat.

**Comment by ruseinov** [comment]
There's a discussion here: https://chainsafeio.slack.com/archives/C01SZBSC6KZ/p1707717786883299.

I appreciate that this is a private link, but still keeping it here should we need to find the thread fast.

**Comment by LesnyRumcajs** [comment]
@ruseinov keep in mind that older release used older crate versions. It might make sense to introduce several flavours of this dependency, just like for FVM.

**Comment by ruseinov** [comment]
> @ruseinov keep in mind that older release used older crate versions. It might make sense to introduce several flavours of this dependency, just like for FVM.

I'm unsure why we should bother with that. Why don't we just update latest fil-actor-states as discussed, release it and integrate into latest forest with TokenStateExt. Older release depended on older crate versions, that is fine. Why several flavours?

**Comment by LesnyRumcajs** [comment]
We should bother with that if there are breaking changes between the crate releases, same reason why we have FVM2, FVM3, and FVM4. It might not be needed now for the token crate, but it's something we should keep in mind.

**Comment by ruseinov** [comment]
> We should bother with that if there are breaking changes between the crate releases, same reason why we have FVM2, FVM3, and FVM4. It might not be needed now for the token crate, but it's something we should keep in mind.

I'm still unsure why. FVM versioning is a really bad example, it's horrible to deal with and I hope we never have to resort to that kind of thing. I'm pretty sure that if we had control over it - it could have been done much nicer, at least the compatibility layer. 

I'm unclear on why we can't introduce a breaking change in a new release of the crate and then if people want to use the old version - they can. In any case this consideration warrants a valid use-case, before this materialises - I think we're safe to proceed in a way that we did so far. 

I would also argue that before resorting to the FVM approach we should carefully weigh all the pros and cons, I'd rather have several codepaths and versioning WITHIN the crate, than having to deal with 3 versions of the same crate and trying fit them into the same product. I know Lotus isn't the best example, but so far they have managed to maintain actors compatibility without resorting to anything as extreme as that.

**Comment by LesnyRumcajs** [comment]
I'm not sure what are you even referring to. Let me rephrase. We may need a dependency like `frc46_token_9 = 9.0.0`, `frc46_token_10 = 10.0.0` and so on, if there are breaking changes between those two. For now, there doesn't seem to be such need. If it's still unclear, let's just drop it.

**Comment by ruseinov** [comment]
> I'm not sure what are you even referring to. Let me rephrase. We may need a dependency like `frc46_token_9 = 9.0.0`, `frc46_token_10 = 10.0.0` and so on, if there are breaking changes between those two. For now, there doesn't seem to be such need. If it's still unclear, let's just drop it.

I wasn't sure what you were referring to, I thought we are talking about `fil-actor-states`. I'm not sure why we are discussing this now as it's currently irrelevant. When it becomes relevant - let's discuss it again.
